os/drivers/input/ist415.c

```diff
@@ -89,185 +94,1899 @@ struct touchscreen_ops_s g_ist415_ops = {
+ * Name: put_unaligned_be32
  ****************************************************************************/
 
-static int ist415_get_touch_data(struct ist415_dev_s *dev, FAR void *buf)
+static __inline void put_unaligned_be32(uint32_t value, void *addr) {
+    uint8_t *p = (uint8_t *)addr;
+
+    p[0] = (value >> 24) & 0xFF; // Most significant byte
+    p[1] = (value >> 16) & 0xFF;
+    p[2] = (value >> 8) & 0xFF;
+    p[3] = value & 0xFF;         // Least significant byte
+}
+
+/****************************************************************************
+ * Name: cpu_to_be32
+ ****************************************************************************/
+
+static __inline uint32_t cpu_to_be32(uint32_t value) {
+    return ((value & 0xFF000000) >> 24) | // Move byte 3 to byte 0
+           ((value & 0x00FF0000) >> 8)  | // Move byte 2 to byte 1
+           ((value & 0x0000FF00) << 8)  | // Move byte 1 to byte 2
+           ((value & 0x000000FF) << 24);  // Move byte 0 to byte 3
+}
+
+uint32_t ist415x_parse_ver(struct ist415_dev_s *dev, int flag, const uint8_t *buf)
+{
+	uint32_t ver = 0;
+	uint32_t *buf32 = (uint32_t *)buf;
+
+	if (flag == FLAG_MAIN)
+		ver = (uint32_t)buf32[(dev->tags.flag_addr + 0xFC) >> 2];
+	else if (flag == FLAG_RELEASE)
+		ver = (uint32_t)buf32[(dev->tags.flag_addr + 0xF0) >> 2];
+	else if (flag == FLAG_TEST)
+		ver = (uint32_t)buf32[(dev->tags.flag_addr + 0xF4) >> 2];    
+	else if (flag == FLAG_CORE)
+		ver = ((uint32_t)buf32[(dev->tags.flag_addr + 0xEC) >> 2] >> 16) & 0xFFFF;
+	else if (flag == FLAG_CONFIG)
+		ver = (uint32_t)buf32[(dev->tags.flag_addr + 0xEC) >> 2] & 0xFFFF;
+	else
+		touchdbg("Parsing ver's flag is not corrent!\n");
+
+	return ver;
+}
+
+/****************************************************************************
+ * Name: ist415_i2c_write
+ ****************************************************************************/
+
+static int ist415_i2c_write(struct ist415_dev_s *dev, uint8_t cmd, uint8_t *data, int size)
 {
 	struct i2c_dev_s *i2c = dev->i2c;
 	struct i2c_config_s config = dev->i2c_config;
-	struct ts_event_coordinate *p_evt_coord;
-	struct touch_sample_s *data = buf;
-	u8 event[EVENT_PACKET_SIZE];
-	u8 touch_point;
-	u8 *touch_event;
-	u8 eid;
-	int reg[2] = {HCOM_GET_EVENT_1ST, 0};
-
-	int ret = i2c_write(i2c, &config, (uint8_t *)reg, 1);
-	if (ret != 1) {
+	int wsize = sizeof(cmd) + size;
+	uint8_t *buf = (uint8_t *)kmm_malloc(wsize);	
+	
+	buf[0] = cmd;
+	if (size > 0) {
+		memcpy(buf + sizeof(cmd), data, size);
+	}
+
+	int ret = i2c_write(i2c, &config, (uint8_t *)&buf, wsize);
+	if (ret != wsize) {
+		kmm_free(buf);
+		touchdbg("ERROR: I2C write failed\n");		
+		return ret;
+	}
+
+	kmm_free(buf);
+
+	return OK;
+}
+
+/****************************************************************************
+ * Name: ist415_power_mode
+ ****************************************************************************/
+static void ist415_power_mode(struct ist415_dev_s *dev, uint8_t data)
```

Comment:
some kind of important operation which can affect to work flow should read i2c again & verify it.